core/src/main/java/io/el/concurrent/DefaultEventLoopGroup.java

```diff
@@ -0,0 +1,186 @@
+package io.el.concurrent;
+
+import static io.el.internal.ObjectUtil.checkPositive;
+
+import io.el.concurrent.EventLoopChooserFactory.EventLoopChooser;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+
+public abstract class DefaultEventLoopGroup implements EventLoopGroup {
+
+  static final long DEFAULT_SHUTDOWN_TIMEOUT = 15;
+  private final List<EventLoop> children;
+  private final EventLoopChooser chooser;
+
+  /***
+   * Create {@link EventLoop} children with size of {@param nThreads},
+   * then add chooser with {@param chooserFactory}
+   */
+  protected DefaultEventLoopGroup(
+      int nThreads,
+      Executor executor,
+      EventLoopChooserFactory chooserFactory
+  ) {
+    checkPositive(nThreads, "nThreads");
+
+    if (executor == null) {
+      executor = new ThreadPerTaskExecutor(this.newDefaultThreadFactory());
+    }
+
+    this.children = new ArrayList<>();
+    for (int i = 0; i < nThreads; i++) {
+      try {
+        this.children.add(this.newChild(executor));
+      } catch (Exception e) {
+        for (int j = 0; j < i; j++) {
+          if (this.children.get(j).shutdownGracefully(DEFAULT_SHUTDOWN_TIMEOUT, TimeUnit.SECONDS)) {
+            EventLoop el = this.children.get(j);
+            try {
+              while (!el.isTerminated()) {
+                el.awaitTermination(Integer.MAX_VALUE, TimeUnit.SECONDS);
+              }
+            } catch (InterruptedException interrupted) {
+              Thread.currentThread().interrupt();
+              break;
+            }
```

Comment:
I think it's better to call `shutdownGracefully()` for all child first then `awaitTermination()` all. Because `shutdownGracefully()` is done in asynchronous way. So call all child shutdown then await all.